Prevent writing to state cache when migrating the database

eserilev opened the pull request:
Added a new flag `update_cache` to `get_state`. When set to false, prevent writing fetched state to the cache

eserilev committed:
add an update_cache flag to get_state to have more granular control over when we write to the cache

michaelsproul commented on beacon_node/beacon_chain/src/beacon_chain.rs:
```diff
@@ -821,7 +821,7 @@ impl<T: BeaconChainTypes> BeaconChain<T> {
             .ok_or(Error::MissingBeaconBlock(block_root))?;
         let state = self
-            .get_state(&block.state_root(), Some(block.slot()))?
+            .get_state(&block.state_root(), Some(block.slot()), true)?
```
Only used in tests so doesn't matter

michaelsproul commented on beacon_node/beacon_chain/src/builder.rs:
```diff
@@ -299,7 +299,11 @@ where
+            .get_state(
+                &genesis_block.state_root(),
+                Some(genesis_block.slot()),
+                true,
```
This is ok as we almost always have the genesis state in the freezer

michaelsproul committed:
State cache tweaks

- add state-cache-headroom flag to control pruning
- prune old epoch boundary states ahead of mid-epoch states
- never prune head block's state
- avoid caching ancestor states unless they are on an epoch boundary

michaelsproul commented:
I'll PR this to `release-v7.0.0` with a `Co-author` tag for you @eserilev
